capstone-app/src/app/modules/history/wiki-results.service.ts

```diff
@@ -17,15 +19,74 @@ export class WikiResultsService {
     'https://en.wikipedia.org/w/api.php?origin=*&action=parse&page=';
   private static readonly URL_END = '&format=json';
 
-  search(query: string): Observable<WikiSearchResult> {
+  search(query: string): Observable<{title: string, history: string, furtherReading: Map<string, string>}> {
     return this.searchHandler
       .getQueryString(query)
       .pipe(
         switchMap((result) =>
           this.http.get<WikiSearchResult>(
             `${WikiResultsService.URL_BEGINNING}${result}${WikiResultsService.URL_END}`
           )
-        )
+        ),map(history => {
+          if (history.parse.text) {
+            const historyResult = this.fixString(history.parse.text['*']);
+            const title = history.parse.title;
+            return {title, ...historyResult};
+          }
+          else {
+            throw new Error("API did not return a valid response");
+          }
+        }),
+        catchError(() => {
+          console.log('API did not return a valid response.');
+          const title = '';
+          const history = '';
+          const furtherReading = new Map<string, string>();
+          return of({title, history, furtherReading});
+        })
       );
   }
+
+  fixString(text: string): {history: string, furtherReading: Map<string, string>} {
+    const firstIndex = text.indexOf('<span class="mw-headline" id="History">History</span>', 0);
+    if (firstIndex !== -1){
+      const firstPartOfString = text.substring(firstIndex, text.length);
+      const endIndex = firstPartOfString.indexOf('<h2>', 0);
+      const startIndex = firstPartOfString.indexOf('</h2>', 0);
+      let middleOfString = firstPartOfString.substring(startIndex, endIndex);
+      const paragraphs = middleOfString.split('<p>');
+      if (paragraphs.length > 6){
+        middleOfString = paragraphs.slice(0,5).join('');
+      }
+      const furtherReading = this.findHrefs(middleOfString);
+      let history = middleOfString.split(/<.*?>/g).join('');
+      history = history.split(/&.*?;/g).join('');
+      return {history, furtherReading};
+    }
+    console.log('The city page was found, but unfortunately there was no history paragraph found!');
+    let history = text.split(/<.*?>/g).join('');
+    history = history.split(/&.*?;/g).join('');
+    return {history, furtherReading: new Map()};
```

Comment:
Is there a reason we're returning an empty map?

Reply:
Yes, we need to return a new map in order to return all the fields. But there is no data to put in the map.

Reply:
Well technically we don't need to return a map either. Returning an empty map was just one of the ways to do this.
When you set the return type you can make furtherReading optional by applying a question mark between the variable name and the colon that comes before the type annotation:

fixString([...]): {history: string, furtherReading?: Map<string, string>} {
[...]
}

or using an interface:

export interface ArticleData {
  history: string;
  furtherReading?: string;
}
[...]
fixString([...]): ArticleData {
}